src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/BrowserWebSockets/BrowserWebSocket.cs

```diff
@@ -0,0 +1,529 @@
+                return _state switch
+                {
+                    _created => WebSocketState.None,
+                    _connecting => WebSocketState.Connecting,
+                    _disposed => WebSocketState.Closed, // We only get here if disposed before connecting
+                    _ => WebSocketState.Closed
+                };
+            }
+        }
+
+        private static WebSocketState ReadyStateToDotNetState(int readyState) =>
+            // https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
+            readyState switch
+            {
+                0 => WebSocketState.Connecting, // 0 (CONNECTING)
+                1 => WebSocketState.Open, // 1 (OPEN)
+                2 => WebSocketState.CloseSent, // 2 (CLOSING)
+                3 => WebSocketState.Closed, // 3 (CLOSED)
+                _ => WebSocketState.None
+            };
+
+        public override WebSocketCloseStatus? CloseStatus => _innerWebSocket == null ? null : _innerWebSocketCloseStatus;
+
+        public override string? CloseStatusDescription => _innerWebSocket == null ? null : _innerWebSocketCloseStatusDescription;
+
+        public override string? SubProtocol => _innerWebSocket != null && !_innerWebSocket.IsDisposed ? _innerWebSocket!.GetObjectProperty("protocol")?.ToString() : null;
+
+        #endregion Properties
+
+        internal async Task ConnectAsyncJavaScript(Uri uri, CancellationToken cancellationToken, List<string>? requestedSubProtocols)
+        {
+            var tcsConnect = new TaskCompletionSource<bool>();
+
+            // For Abort/Dispose.  Calling Abort on the request at any point will close the connection.
+            _cts.Token.Register(AbortRequest);
+
+            // Wrap the cancellationToken in a using so that it can be disposed of whether
+            // we successfully connected or failed trying.
+            // Otherwise any timeout/cancellation would apply to the full session.
+            // In the failure case we need to release the references and dispose of the objects.
+            using (cancellationToken.Register(() => tcsConnect.TrySetCanceled()))
+            {
+                try
+                {
+                    JavaScript.Array? subProtocols = null;
+                    if (requestedSubProtocols?.Count > 0)
+                    {
+                        subProtocols = new JavaScript.Array();
+                        foreach (var item in requestedSubProtocols)
+                        {
+                            subProtocols.Push(item);
+                        }
+                        _innerWebSocket = new HostObject("WebSocket", uri.ToString(), subProtocols);
+                    }
+                    else
+                    {
+                        _innerWebSocket = new HostObject("WebSocket", uri.ToString());
+                    }
+                    _innerWebSocket.SetObjectProperty("binaryType", "arraybuffer");
+
+                    subProtocols?.Dispose();
```

Comment:
This can now also go into the `requestedSubProtocols?.Count > 0` case (maybe as a `using`) because it is only used there.

Reply:
yes agree.  Thanks